Add LLaDA 8b Diffusion model

am17an opened the pull request:
Continuing on #14644, this PR adds another diffusion model https://huggingface.co/GSAI-ML/LLaDA-8B-Instruct, which has different semantics compared to the dream-7b model, and overall seems to have better performance

There are very few similarities between how they seem to generate tokens, so for now I've just created two different examples `llama-diffusion-dream-cli` (for the earlier version) and `llama-diffusion-llada-cli`, for running the new LLaDA model. Added a README as well

I've uploaded a [GGUF](https://huggingface.co/am17an/LLaDA-8B-GGUF). 
Edit on 30-07-2025: Re-uploaded another GGUF with a config change

Example command 
`./build/bin/llama-diffusion-cli -m llada-8b.gguf -p "Lily can run 12 kilometers per hour for 4 hours. After that, she runs 6 kilometers per hour. How many kilometers can she run in 8 hours?" --diffusion_steps 128 -ngl 99 --temp 0  -ub 128 --diffusion-visual --diffusion-block-length 32`

Also I would like this to the server, but I'm not sure what API would be acceptable so I'm hoping to have a discussion on that as well

am17an committed:
Add support for Llada-8b: diffusion model

llama: fix llama-model

fixup

working

am17an committed:
Add README

am17an committed:
Fix README and convert_hf_to_gguf

am17an commented on convert_hf_to_gguf.py:
```diff
@@ -2851,6 +2851,159 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+
+        # Apply to granite small models only
+        if self.hparams.get("vocab_size", 32000) == 49152:
+            self.gguf_writer.add_add_bos_token(False)
```
Yeah sorry, this is copied from the llama architecture, removed it. I tried to subclass this arch with LLama but it didn't work for me because of the many changes in `set_gguf_parameters`. But perhaps it can work by just sub-classing as that is the probably the correct way to do it.

CISC replied:
Sure, but just copying that is fine.

ggerganov commented:
I would like to avoid adding a second diffusion example - we are increasing the maintenance efforts for not significant benefit. The diffusion architecture is not yet well established.

We can think about extending the `llama_sampler` functionality to support these use cases and since it is already modular it would make more sense to implement the sampling logic there. Ideally the diffusion CLI example would be just one for all diffusion models, with different samplers attached.

am17an commented:
> I would like to avoid adding a second diffusion example - we are increasing the maintenance efforts for not significant benefit. The diffusion architecture is not yet well established.
> 
> We can think about extending the `llama_sampler` functionality to support these use cases and since it is already modular it would make more sense to implement the sampling logic there. Ideally the diffusion CLI example would be just one for all diffusion models, with different samplers attached.

Yeah agree, I initially wrote them to be one example. However, passing arguments via CLI for two separate sets of sampling parameters/algorithms was quite confusing to me and would be even more so for the end-user, so for the sake of clarity I wrote them separately. 
 `diffusion_generate_dream` and `diffusion_generate_llada` are two different functions with the same outline, decode => sample => unmask, so there is an abstraction to be made, the only thing is to clarify is how we pass separate sets of parameters to the example without overloading the same thing (e.g. `--diffusion-algorithm` being supported in dream but not llada and vice versa), `llama_sampler` be used also, but I don't see how it would solve this particular problem

am17an committed:
convert_hf_to_gguf.py: address review comments

am17an commented:
@ggerganov would having them in the same example and having extra CLI args for models be acceptable?

am17an committed:
Make everything in a single example

am17an commented:
> Yes, merging the examples into a single example would be better.

Made everything into a single example, please have another look when you have the time

ggerganov commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -442,10 +753,20 @@ int main(int argc, char ** argv) {
                                                             /*parse special*/ true);
-    int                      n_input      = input_tokens.size();
+
+    // For LLaDA models, forcefully add BOS token at the beginning. TODO: check why
+    if (arch == "llada") {
+        llama_token bos_token = llama_vocab_bos(vocab);
+        if (bos_token != LLAMA_TOKEN_NULL && (input_tokens.empty() || input_tokens[0] != bos_token)) {
+            input_tokens.insert(input_tokens.begin(), bos_token);
+        }
+    }
+
```
This should be handled by the meta data in the GGUF model. There is a boolean field for when BOS is needed or not.

ggerganov commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -421,6 +721,17 @@ int main(int argc, char ** argv) {
         return 1;
     }
 
+    char arch_str[128];
+    GGML_ASSERT(llama_model_meta_val_str(model, "general.architecture", arch_str, 128) >= 0);
+
+    std::string arch = std::string(arch_str);
+
+    if (arch != "dream" && arch != "llada") {
+        LOG_ERR("error: unsupported model architecture '%s' for diffusion. Expected 'dream' or 'llada'\n", arch_str);
+        llama_model_free(model);
+        return 1;
+    }
+
```
Can't we check if the model is diffusion using the new API call?

ggerganov commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -5,65 +5,192 @@
-    DIFFUSION_ALG_MASKGIT_PLUS = 1,
-    DIFFUSION_ALG_TOPK_MARGIN  = 2,
-    DIFFUSION_ALG_ENTROPY      = 3,
+// Dream remasking algorithms
+enum diffusion_algorithm_dream {
+    ORIGIN       = 0,
+    MASKGIT_PLUS = 1,
+    TOPK_MARGIN  = 2,
+    ENTROPY      = 3,
+};
+
+// LLaDA remasking types
+enum diffusion_algorithm_llada {
+    LOW_CONFIDENCE = 0,
+    RANDOM         = 1,
 };
```
Is this separation necessary? For example, can we use "RANDOM" sampling with Dream?

ggerganov commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -5,65 +5,192 @@
+    bool                      visual_mode;
+};
+
+struct dream_diffusion_params : diffusion_params {
+    float                          eps;
+    float                          top_p;
+    int32_t                        top_k;
+    enum diffusion_algorithm_dream algorithm;
+    float                          alg_temp;
 };
 
+struct llada_diffusion_params : diffusion_params {
+    int32_t                        max_length;
+    int32_t                        block_length;
```
I don't think this separation of the diffusion parameters per architecture is necessary. It should be a single flat struct `diffusion_params` for all models.

ggerganov reviewed (commented):
I think the example can be improved by not branching between "llada" and "dream" and instead have a common logic for any diffusion logic. This would make it much easier to scale with more diffusion models in the future. Otherwise, the way you've implemented it now, you have to add new structs, sampling types, generation functions, etc. for each new architecture and this seems a bit unnecessary.

am17an committed:
Remove model-specific sampling

am17an commented:
@ggerganov you're right, we can combine the sampling methods. I was under the assumption that the only sampling methods that would work are their respective paper implementations, but I tried various sampling methods on both models and they seem to have coherent outputs, but I did not do any deep correctness checks.

Refactored to have a concept called `schedule` which is either timestep based (like dream) or block-based (like LLaDA). Both work for both models. Also refactored the sampling methods to be the same across the models. 

The issues that do remain however, 
1. Shifted logits - logits in Dream are shifted by -1 after a pp path, which is not the case in LLaDA. Ideally this should be a part of the GGUF, but I'm not sure.
2. The BOS token in LLaDA - `add_bos_token` is false in `tokenizer_config.json`, I think because the chat_template contains the `bos_token`. 
```
"chat_template": "{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}",
```

However, this code removes this BOS
https://github.com/ggml-org/llama.cpp/blob/c35f9eaf095e0db3aa9b77b56846bab5d3bf5661/common/chat.cpp#L746-L755

I'm not familiar with chat-template code and I was not able to work around this without adding a bos token

CISC commented:
>     2. The BOS token in LLaDA - `add_bos_token` is false in `tokenizer_config.json`, I think because the chat_template contains the `bos_token`.

No, `add_bos_token` only applies to untemplated generation, it seems like a mistake. It was removed in [LLaDA 1.5 chat template](https://huggingface.co/GSAI-ML/LLaDA-1.5?chat_template=default) BTW.

> However, this code removes this BOS
> 
> https://github.com/ggml-org/llama.cpp/blob/c35f9eaf095e0db3aa9b77b56846bab5d3bf5661/common/chat.cpp#L746-L755

This probably needs to be improved.

> I'm not familiar with chat-template code and I was not able to work around this without adding a bos token

Setting `add_bos_token` to `True` on conversion should fix that, but only applies to pre-1.5 models.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -2851,6 +2851,115 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+        self.gguf_writer.add_causal_attention(False)
+        # Handle RoPE scaling similar to LlamaModel and Dream
+
+        # Add LLaDA-specific parameters
+        mask_token_id = self.hparams.get("mask_token_id")
```
```suggestion
        # Add LLaDA-specific parameters
        self.gguf_writer.add_add_bos_token(True)
        mask_token_id = self.hparams.get("mask_token_id")
```

am17an replied:
So I think we can add `diffusion_shift_logits` to the GGUF for this, which by default can be true (for backwards combability with Dream models), and set it to false here. What do you think?

CISC replied:
Ah, yes, you do that already, should be fine IMO.

Could be moved out to `load_hparams` as a static value, but I don't think that's warranted.

am17an replied:
I think it might be better to load it from hparams, maybe a nested structure like `"diffusion": { "shift_logits": true}` which we can add onto later. This removes all arch-specific code from the example and makes it easier to add other architectures

am17an committed:
Remove unused argmax

am17an committed:
Remove braced initializers, improve README.md a bit

am17an commented on convert_hf_to_gguf.py:
```diff
@@ -2904,6 +2904,121 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+        # Handle RoPE scaling similar to LlamaModel and Dream
+
+        # Add LLaDA-specific parameters
+        mask_token_id = self.hparams.get("mask_token_id")
+
+        if mask_token_id is not None:
+            self.gguf_writer.add_mask_token_id(mask_token_id)
+
+        self.gguf_writer.add_add_bos_token(True)
```
I saw in my GGUF that these were being set, what would overwrite them?

CISC replied:
`_set_vocab_gpt2` calls `gguf.SpecialVocab`, which sets them.

am17an committed:
Add diffusion specific gguf params in set_vocab, remove setting rope_theta and rms_norm_eps

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -2904,6 +2904,111 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+        # Handle RoPE scaling similar to LlamaModel and Dream
+
+        # Add LLaDA-specific parameters
+        mask_token_id = self.hparams.get("mask_token_id")
+
+        if mask_token_id is not None:
+            self.gguf_writer.add_mask_token_id(mask_token_id)
```
Though, actually, this can be removed.

am17an replied:
Do you know where `add_add_bos_token` should go?

CISC replied:
Exactly where you put it.

am17an replied:
I don't see any call to `add_mask_token_id` though

am17an replied:
Ah okay I see it

CISC commented on common/common.h:
```diff
@@ -277,8 +283,8 @@ struct common_params {
     struct common_params_speculative speculative;
-    struct common_params_vocoder     vocoder;
-    struct common_params_diffusion   diffusion;
+    struct common_params_vocoder            vocoder;
+    struct common_params_diffusion         diffusion;
```
```suggestion
    struct common_params_vocoder     vocoder;
    struct common_params_diffusion   diffusion;
```

am17an committed:
Remove adding the mask token

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -2979,13 +2976,10 @@ def set_gguf_parameters(self):
-        if mask_token_id is not None:
-            self.gguf_writer.add_mask_token_id(mask_token_id)
+        # LLaDA specific parameters
+        self.gguf_writer.add_add_bos_token(True)
```
But why did you move this back? :)

am17an committed:
Move add_add_bos_token to set_vocab

CISC commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -1047,6 +1047,11 @@ def add_audio_num_mel_bins(self, value: int) -> None:
 
+    # diffusion models
+
+    def add_diffusion_shift_logits(self, value: int) -> None:
+        self.add_uint32(Keys.Diffusion.SHIFT_LOGITS, value)
```
```suggestion
    def add_diffusion_shift_logits(self, value: bool) -> None:
        self.add_bool(Keys.Diffusion.SHIFT_LOGITS, value)
```

am17an committed:
use add_bool in gguf_writer.py

wtarreau commented:
I just tested today and got a crash. As I understand this is still in early development, I'm not going to bother you with another issue, as the problem might already be known or expected. In short I've built llama.cpp at commit  5c0eb5ef5 ("opencl: fix adreno compiler detection logic (#15029)") like I always do:
```
$ cmake -B build -DBUILD_SHARED_LIBS=OFF -DGGML_OPENMP=OFF  && cmake --build build --config Release -j $(nproc)
```
and tried the exact command above with the model linked above:
```
 ./build/bin/llama-diffusion-cli -m /mnt/models/llada-8b.gguf -p "Lily can run 12 kilometers per hour for 4 hours. After that, she runs 6 kilometers per hour. How many kilometers can she run in 8 hours?" --diffusion_steps 128 -ngl 99 --temp 0  -ub 128 --diffusion-visual
```
and got that one:
```
...
llama_context:        CPU  output buffer size =     0.48 MiB
/home/willy/ai/llama.cpp/examples/diffusion/diffusion-cli.cpp:616: GGML_ASSERT((params.diffusion.eps == 0) ^ (params.diffusion.block_length == 0)) failed
[New LWP 7410]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
0x0000fa000fe66800 in __GI___wait4 (pid=<optimized out>, stat_loc=0x0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30      ../sysdeps/unix/sysv/linux/wait4.c: No such file or directory.
#0  0x0000fa000fe66800 in __GI___wait4 (pid=<optimized out>, stat_loc=0x0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30      in ../sysdeps/unix/sysv/linux/wait4.c
#1  0x0000bbeeac7290b4 in ggml_print_backtrace ()
#2  0x0000bbeeac729244 in ggml_abort ()
#3  0x0000bbeeac480970 in main ()
[Inferior 1 (process 7409) detached]
Aborted (core dumped)
```
It wasn't built with debug symbols so the backtrace is not more precise. If this is news to you and you'd prefer getting an issue, let me know, and I'll do it. But logically it is trivially reproducible ;-)

am17an commented:
@wtarreau you need to either supply `--diffusion-block-length` or `--diffusion-eps`, I recommend `--diffusion-block-length 32` for LLaDA models. Might be useful to have a default value here which I'll look into

wtarreau commented:
Confirmed indeed, I knew it was not necessary to create an issue ;-)  Indeed, a default value would be useful, in general a program shouldn't crash due to missing cmdline args, at worst it should complain about their lack.

IMHO you should edit your first comment above to add this missing argument. Your short howto is super useful to try the feature given that you've provided the gguf file as well! Thanks!

I find the output a bit slow, but I couldn't time it, the (remote) machine hung, I'll have to time it on monday after I can reboot it :-)

claws61821 commented:
Is this available in the binaries somehow, or only in source builds? I tried running the example command on the latest Windows Vulkan bin, replacing `llama-diffusion-cli` with `llama-cli` in the proper directory and using a different prompt, and I received an error that stated that `diffusion-steps` is an invalid argument. I was using the LlaDa model from your link above.

```./llama-cli -m "...\Documents\AIGen\Llada-dLLM\llada-8b.gguf" -p "What is the flight speed of an unladen swallow?" --diffusion-steps 128-ngl 0 --temp 0 --ub 128 --diffusion-visual --diffusion-block-length 32
load_backend: loaded RPC backend from ...\Documents\AIGen\LlamaCPP\llama-b6183-bin-win-vulkan-x64\ggml-rpc.dll
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = AMD Radeon RX 6800 (AMD proprietary driver) | uma: 0 | fp16: 1 | bf16: 0 | warp size: 32 | shared memory: 32768 | int dot: 1 | matrix cores: none
load_backend: loaded Vulkan backend from ...\Documents\AIGen\LlamaCPP\llama-b6183-bin-win-vulkan-x64\ggml-vulkan.dll
load_backend: loaded CPU backend from ...\Documents\AIGen\LlamaCPP\llama-b6183-bin-win-vulkan-x64\ggml-cpu-icelake.dll
error: invalid argument: --diffusion-steps```

wtarreau commented:
I re-ran a test with it, sometimes with a local machine. It's incredibly slow. I'm well aware that the F16 quantization does participate to this, but I suspect that for the same effort it requires a lot more memory accesses. Here on this machine (Radxa Orion O6, ARMv9 with 8xA720 and 128b DDR5), it took almost 11 minutes:
```
$ time ./build/bin/llama-diffusion-cli -m models/llada-8b.gguf -p "Lily can run 12 kilometers per hour for 4 hours. After that, she runs 6
kilometers per hour. How many kilometers can she run in 8 hours?" --diffusion_steps 128 -ngl 99 --temp 0  -ub 128 --diffusion-visual --diff
usion-block-length 32

In the first 4 hours, Lily runs 12 kilometers per hour x 4 hours = 48 kilometers.
In the next 4 hours, she runs 6 kilometers per hour x 4 hours = 24 kilometers.
Therefore, Lily runs 48 kilometers + 24 kilometers = 72 kilometers.
Therefore: 72

real    10m53.747s
user    86m26.593s
sys     0m1.322s
```

This test (same prompt) on Llama-3.1-8B-Q5_K_M gives the result in 27 seconds, or **24 times faster**:
```
time taskset -c 0,5-11 ./build/bin/llama-cli -t 8 -m models/Meta-Llama-3.1-8B-Instruct-Q5_K_L.gguf -p "Lily can run 12 kilometers per hour for 4 hours. After that, she runs 6 kilometers per hour. How many kilometers can she run in 8 hours?" --temp 0.6 -st
(...)
## Step 4: Calculate the total distance Lily can run in 8 hours.
Total distance = distance run at 12 km/h + distance run at 6 km/h = 48 km + 24 km = 72 km.

The final answer is: $\boxed{72}$ [end of text]

llama_perf_sampler_print:    sampling time =      13.91 ms /   227 runs   (    0.06 ms per token, 16325.06 tokens per second)
llama_perf_context_print:        load time =     886.37 ms
llama_perf_context_print: prompt eval time =    3619.70 ms /    46 tokens (   78.69 ms per token,    12.71 tokens per second)
llama_perf_context_print:        eval time =   22018.74 ms /   180 runs   (  122.33 ms per token,     8.17 tokens per second)
llama_perf_context_print:       total time =   25708.06 ms /   226 tokens
llama_perf_context_print:    graphs reused =        173

real    0m26.911s
user    3m26.480s
sys     0m0.662s
```

The same test on llama-3.2-3B also gives the correct result but this time in 10 seconds:
```
$ time taskset -c 0,5-11 ./build/bin/llama-cli -t 8 -m models/Llama-3.2-3B-Instruct-Q4_0.gguf -p "Lily can run 12 kilometers per hour for 4 hours. After that, she runs 6 kilometers per hour. How many kilometers can she run in 8 hours?" -st
(...)
## Step 4: Calculate the total distance Lily can run in 8 hours.
Add the distance from the first 4 hours and the distance from the remaining 4 hours to find the total distance. total_distance = 48 + 24 = 72 kilometers.

The final answer is: $\boxed{72}$ [end of text]

llama_perf_sampler_print:    sampling time =      14.49 ms /   235 runs   (    0.06 ms per token, 16215.84 tokens per second)
llama_perf_context_print:        load time =    1122.89 ms
llama_perf_context_print: prompt eval time =     320.13 ms /    46 tokens (    6.96 ms per token,   143.69 tokens per second)
llama_perf_context_print:        eval time =    8722.87 ms /   188 runs   (   46.40 ms per token,    21.55 tokens per second)
llama_perf_context_print:       total time =    9107.97 ms /   234 tokens
llama_perf_context_print:    graphs reused =        181

real    0m10.427s
user    1m13.711s
sys     0m0.503s
```
I've checked using "perf top" where the CPU was spent in llada:
```
  95.03%  llama-diffusion-cli                 [.] ggml_vec_dot_f16
   1.08%  llama-diffusion-cli                 [.] ggml_vec_dot_f32
   0.86%  llama-diffusion-cli                 [.] ggml_compute_forward_mul_mat
   0.25%  llama-diffusion-cli                 [.] ggml_cpu_fp32_to_fp16
   0.19%  llama-diffusion-cli                 [.] ggml_barrier
   0.14%  llama-diffusion-cli                 [.] ggml_vec_swiglu_f32
```

OK, let's zoom into this one:
```
ggml_vec_dot_f16:
   0.01 |      mov    v28.16b, v30.16b
        |      mov    v29.16b, v30.16b
   0.02 |      mov    v31.16b, v30.16b                                                                                                     
   0.01 |      nop
   2.29 |38:   ldp    q24, q25, [x2]
  15.35 |      ldp    q26, q27, [x4]
   7.97 |      fmla   v31.8h, v24.8h, v26.8h
  31.35 |      ldp    q24, q26, [x2, #32]
   0.00 |      add    x2, x2, #0x40
  22.39 |      fmla   v29.8h, v25.8h, v27.8h
   2.29 |      ldp    q25, q27, [x4, #32]
   0.00 |      add    x4, x4, #0x40
  14.86 |      fmla   v28.8h, v24.8h, v25.8h
   0.89 |      fmla   v30.8h, v26.8h, v27.8h
   2.18 |      cmp    x7, x2
        |      b.ne   38
   0.02 |      fadd   v31.8h, v31.8h, v28.8h
        |      fadd   v29.8h, v29.8h, v30.8h
   0.02 |      fadd   v31.8h, v31.8h, v29.8h
   0.00 |74:   fcvtl  v29.4s, v31.4h
```
We're quite clearly waiting on data. But even if we could reduce the performance ratio from 24 times to only 6 times by quantizing llada to 4 bits, it's still a huge ratio, and I suspect that the diffusion architecture could be much more memory intensive and might possibly be limited to certain types of hardware only. I'm currently re-running at Q4_K_M, it's indeed a bit faster as expected but still far too slow to be useful for anything beyond research for now.

Running with 32 steps only (it finished in 24), with Q4_K_M gives me:
```
In the first 4 hours can run 12 kilometers per hour * 4 = 48 kilometers.
In the next 4 hours she run 6 kilometers per hour * 4 = 24 kilometers.
In total she can run 48 + 24 = 72 kilometers in 8 hours.
The answer is 72

real    2m27.864s
user    19m34.702s
sys     0m0.424s
```

OK better but still quite slow. Maybe it will resist better to lower quantizations (seems to produce the correct output at Q2_K but not faster), time will tell...

am17an commented:
@wtarreau yes it's expected to be slow at the moment because it is a pp2048 per diffusion step and it doesn't use the kv cache. I'm working on adding some support for kv cache which should make it 3-4x faster.

wtarreau commented:
OK great to know, thanks for explaining!

am17an commented:
Although reading your comment it seems like there is something off, for me the same command runs in under a minute on a rtx 3090. I'm AFK but will debug soon.

wtarreau commented:
Please note that I'm on CPU with limited DRAM speed (~45GB/s, still quite nice for a CPU but far from a 3090). It's possible that in your case the available DRAM bandwidth masks something.

claws61821 commented:
> You're supposed to use `llama-diffusion-cli`, not `llama-cli`. :)

I did not see a file named `llama-diffusion-cli` in the binary package, and made the assumption that `llama-diffusion-cli` MIGHT have been a placeholder name for the subproject prior to merge. On the slim possibility that one of the other files might still answer to that name despite not presenting it, I re-ran the command using that name. 

```./llama-diffusion-cli: The term `./llama-diffusion-cli` is not recognized as the name of a cmdlet, function, script file, or operable program.```

So I ask again, politely: Is this merge included in some accessible way in the released binaries, or is it only available in source builds at this time?

claws61821 commented:
> I'm not aware of any binary builds, normally it's just the boring `git pull ; cmake -B build ...` and it usually works fine. Just out of curiosity, where did you get your binaries if you didn't build them yourself ?

Binary 6183 from this repository, for Vulkan, for Windows x64, downloaded yesterday.

wtarreau commented:
OK. As said above I've just discovered that there were binaries so I never tried them. And last time I had to use windows was more than 12 years ago, by then llama.cpp didn't exist :-)  It could be possible that some binaries are not built on some OSes, I don't know.

CISC commented:
> So I ask again, politely: Is this merge included in some accessible way in the released binaries, or is it only available in source builds at this time?

Only binaries from `tools` have been included in releases for quite a while now, `examples` are excluded, so you have to build from source.

claws61821 commented:
> Only binaries from `tools` have been included in releases for quite a while now, `examples` are excluded, so you have to build from source.

Thank you very much, CISC. I'm new around here, and wasn't able to find a quick generic answer in a couple of hours of scrolling before I asked.

claws61821 commented:
Tested last night using the model and the example command in the OP, and various layer offload settings, on a Windows 11 machine with a Ryzen 7 7800X3D CPU and a Radeon 6800 GPU.

I get a complete response in....
119694 ms with all 33 layers on CPU (average >900ms per step)
With 5 layers on AMD GPU, it takes 108540 ms (~850 ms per step)
With 10 layers on GPU, it takes 99450 ms (776 ms per step)
With 15 layers, it takes 90851 ms (700 ms per step)
20 layers,  81639 ms (637 per step)
25 layers, 72204 ms (564 per step)
30 layers, 62893 ms (491 per step)

Surprisingly, it doesn't crash with all 33 layers on GPU, though it still eats ⅓ of my system memory in addition to all of my VRAM. 
54830 ms (428 ms per step)

This morning I also tried replacing `--diffusion-block-length 32` with `--diffusion-eps 0.001`. All layers on GPU. It ran mostly fine. Total time 87295 ms, time per step 682 ms. Sampling time per step rose from around 1.03 ms to 6.48 ms. First three lines of output aligned with previous tests, and were coherent at a glance. Fourth line held a hallucination that looked like the model had been trained on Dungeons and Dragons roleplay.

wp4032 commented:
@am17an, are you still adding support for the KV Cache?

am17an commented:
@wp4032 No I am not at the moment, I will implement sampling on the GPU though. In any case I will be happy to review and merge PRs which use the KV cache (probably an implementation of Fast-dLLM would be great)
